src/Data/BitVector/Sized/Internal.hs

```diff
@@ -1,419 +1,748 @@
--- from the type context. The input (an unbounded data type, hence with an
--- infinite-width bit representation), whether positive or negative, is silently
--- truncated to fit into the number of bits demanded by the return type.
+-- | Bitvector datatype, parameterized by width.
+data BV (w :: Nat) :: * where
+  -- | We store the value as an 'Integer' rather than a 'Natural',
+  -- since many of the operations on bitvectors rely on a two's
+  -- complement representation. However, an invariant on the value is
+  -- that it must always be positive.
+  BV :: Integer -> BV w
+  deriving (Generic, Show, Read, Eq, Ord, Lift)
+
+instance ShowF BV
+
+instance EqF BV where
+  BV bv `eqF` BV bv' = bv == bv'
+
+instance Hashable (BV w) where
+  hashWithSalt salt (BV i) = hashWithSalt salt i
+
+instance HashableF BV where
+  hashWithSaltF salt (BV i) = hashWithSalt salt i
+
+----------------------------------------
+-- BV construction
+-- | Internal function for masking the input integer *without*
+-- checking that the width is representable as an 'Int'. We use this
+-- instead of 'mkBV' whenever we already have some guarantee that the
+-- width is legal.
+mkBV' :: NatRepr w -> Integer -> BV w
+mkBV' w x = BV (P.toUnsigned w x)
+
+-- | Construct a bitvector with a particular width, where the width is
+-- provided as an explicit `NatRepr` argument. The input 'Integer',
+-- whether positive or negative, is silently truncated to fit into the
+-- number of bits demanded by the return type. The width cannot be
+-- arbitrarily large; it must be representable as an 'Int'.
+--
+-- >>> mkBV (knownNat @4) 10
+-- BV 10
+-- >>> mkBV (knownNat @2) 10
+-- BV 2
+-- >>> mkBV (knownNat @4) (-2)
+-- BV 14
+mkBV :: NatRepr w
+     -- ^ Desired bitvector width
+     -> Integer
+     -- ^ Integer value to truncate to bitvector width
+     -> BV w
+mkBV w x = checkNatRepr w $ mkBV' w x
+
+-- | Panic if an unsigned 'Integer' does not fit in the required
+-- number of bits, otherwise return input.
+checkUnsigned :: NatRepr w
+              -> Integer
+              -> Maybe Integer
+checkUnsigned w i = if i < 0 || i > P.maxUnsigned w
+  then Nothing
+  else Just i
+
+-- | Like 'mkBV', but returns 'Nothing' if unsigned input integer cannot be
+-- represented in @w@ bits.
+mkBVUnsigned :: NatRepr w
+             -- ^ Desired bitvector width
+             -> Integer
+             -- ^ Integer value
+             -> Maybe (BV w)
+mkBVUnsigned w x = checkNatRepr w $ BV <$> checkUnsigned w x
+
+-- | Panic if a signed 'Integer' does not fit in the required number
+-- of bits. Returns an unsigned positive integer that fits in @w@
+-- bits.
+signedToUnsigned :: 1 <= w => NatRepr w
+                 -- ^ Width of output
+                 -> Integer
+                 -> Maybe Integer
+signedToUnsigned w i = if i < P.minSigned w || i > P.maxSigned w
+  then Nothing
+  else Just $ if i < 0 then i + P.maxUnsigned w + 1 else i
```

Comment:
Doesn't seem to actually panic as the comment indicates

Reply:
Thanks -- comment is out of date